Fix for the EntityItemPickupEvent

Draco18s opened the pull request:
Clones the item entity prior to allowing the player to pick it up, allowing the post-added-to-inventory event supply data about the item that was picked up (which is also cloned), rather than TileAir x0 (due to the original being an empty stack). Addresses issue #3419

Draco18s committed:
Fix for the EntityItemPickupEvent

Clones the item entity prior to allowing the player to pick it up,
letting the post-added-to-inventory event supply data about the item
that was picked up, rather than TileAir x0

liach commented on src/test/java/net/minecraftforge/debug/PlayerItemPickupEventDebug.java:
```diff
@@ -0,0 +1,38 @@
+public class PlayerItemPickupEventDebug
+{
+
+    private static final boolean ENABLED = true;
```
You should disable this since almost all testers will pick up drops when testing in a world, and the event listener would be very disturbing.

Draco18s replied:
I did not know if it should be "enabled" or "disabled" when submitted to the PR.
For example, OnItemUseFirstTest is still enabled, while NoBedSleepingTest does not have an enabled/disabled flag, while ItemFishedTest is disabled. There's no consistency.

Oh, LivingDamageEventTest is also lacking a enable flag, and bats fly into lava on world-spawn. I got eight of these in the span of 3 seconds:

[11:43:42] [Server thread/INFO] [forgedebuglivingdamage]: Entity EntityBat['Bat'/152, l='New World', x=-173.32, y=10.89, z=40.43] damage from inFire (pre-reduction): 1.0

liach replied:
Then it's fine.

bs2609 replied:
`LivingDamageEventTest` should have a flag to disable it, the log spam is unhelpful.
Generally, event-based test mods should ship disabled by default.

Draco18s replied:
@bs2609 I definitely agree that the log spam is unhelpful, just that I could find no indications as to how I should submit it. I'll go ahead and change it to disabled.

mezz replied:
The test mods that are left enabled by default are generally ones that only work in very specific circumstances, like hopping on a certain block or using a specific item. Anything that might interfere with other tests or anything that creates log messages in normal situations should be disabled.

Draco18s replied:
Something something `LivingDamageEventTest`

mezz replied:
Forge is flawless shh

liach commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -782,15 +782,15 @@
          public Item func_150995_f()
          {
              if (this == WOOD)
-@@ -1057,5 +1763,21 @@
```
You probably should not include this patch change.

Draco18s replied:
Ah ha, you located The Thing that made all the line numbers change. I was looking for that and couldn't find it. I know I didn't touch the file myself, I assumed that it was still "needed" in some capacity but I couldn't figure out why. I can remove it (I presume that the decompile process added that annotation?).

Ah, no, that annotation is in the original, my version just changes line numbers, and I *still* have no idea why.

I can still remove the file though, not a problem.

williewillus replied:
yeah genPatches sometimes changes patches for files you didn't actually change - in that case simply revert it using `git checkout -- <offending patch>` before committing.

Draco18s replied:
I keep the local repo and the dev environment separate from each other and use a file diff utility to move changes. It lets me firewall things so I don't have to use git commands like that often/at all. For fixing it, I just have to checkout the trunk, copy the patch file "somewhere safe" then checkout my branch, and copy it back (or in this case, I opened it in Notepad++ and then told it not to reload from disk, then resaved the file).

Draco18s committed:
Revert Item.java.patch back to truck version

Draco18s committed:
Test mod disabled by default

LexManos commented:
Would it be worth checking the original stack and subtracting its remaining size from the copy so your stack in the event has the count of the items you picked up?

diesieben07 commented:
That sounds like a reasonable idea to me.

Draco18s commented:
Alternatively I could include the original entity again.

Draco18s commented:
Interesting, found a [vanilla bug](https://bugs.mojang.com/browse/MC-122737) that made this more entertaining to do: because the if statement is only true (and therefor fires the event) when the entity on the ground *has no more items* (i.e. the stack was completely absorbed into the player's inventory) the original entity has no value (it will always only contain air and we miss pickups that use partial stacks--hence the vanilla bug: partial stacks aren't tracked by statistics!).

I'm going to tweak the if statement so that it triggers if *any* items are picked up and move the entityIn.onItemPickup call inside the if(stack.isempty) check. Not doing so leaves a phantom entity around (exists on the server, does not exist on the client), but means no animation or sound plays.

Its the only sensible option until the vanilla bug is fixed.

Draco18s committed:
Altered event to contain all relevant information

Item pickup behavior altered to always fire the event and update
statistics, even if the entire entity stack is not picked up.

LexManos commented on src/main/java/net/minecraftforge/fml/common/gameevent/PlayerEvent.java:
```diff
@@ -38,11 +38,13 @@ private PlayerEvent(EntityPlayer player)
-        public final EntityItem pickedUp;
-        public ItemPickupEvent(EntityPlayer player, EntityItem pickedUp)
+        public final EntityItem originalEntity;
+        public final ItemStack pickedUp;
```
getters

Draco18s replied:
Can-do. I will note that...
ItemCraftedEvent, ItemSmeltedEvent, and PlayerChangedDimensionEvent don't use getters.
PlayerRespawnEvent is the only one that does.
And that's just within PlayerEvent (which also does not use a getter for the player).

LexManos replied:
Yes, a lot of the old events havent been updated to use getters.
We are however changing to use getters. So anything new should get them.

Draco18s replied:
Gotcha. Fixed things again; being distracted means I forgot to change the usage portion in the test mod.

Draco18s committed:
Private with getters

Draco18s committed:
Updated the test mod too

Also made the getter name and the field name match more sensibly

ShetiPhian commented on src/main/java/net/minecraftforge/fml/common/gameevent/PlayerEvent.java:
```diff
@@ -35,11 +35,27 @@ private PlayerEvent(EntityPlayer player)
     }
 
     public static class ItemPickupEvent extends PlayerEvent {
-        public final EntityItem pickedUp;
```
This broke binary compatibility.
